modules/nrfxlib/nrf_802154/sl/platform/nrf_802154_platform_sl_lptimer_zephyr.c

```diff
@@ -201,3 +256,103 @@ uint64_t nrf_802154_platform_sl_lptimer_sync_lpticks_get(void)
+	}
+
+	/* Allocate the channel if not already done. */
+	if (m_hw_task.chan == RTC_CHAN_INVALID) {
+		/* The channel allocation cannot take place during module
+		 * initialization, because for nRF53 it would run
+		 * out of resources - some other module temporarily needs
+		 * an rtc channel to initialize. For this reason, this is where
+		 * the allocation is made.
+		 * Once a channel has been successfully allocated, it will be held
+		 * until nrf_802154_platform_sl_lp_timer_deinit is called.
+		 */
+		m_hw_task.chan = z_nrf_rtc_timer_chan_alloc();
+		if (m_hw_task.chan < 0) {
+			m_hw_task.chan = RTC_CHAN_INVALID;
+			hw_task_state_set(HW_TASK_STATE_SETTING_UP, HW_TASK_STATE_IDLE);
+			return NRF_802154_SL_LPTIMER_PLATFORM_NO_RESOURCES;
+		}
+	}
+
+	if (z_nrf_rtc_timer_set(m_hw_task.chan, fire_lpticks, NULL, NULL) != 0) {
+		hw_task_state_set(HW_TASK_STATE_SETTING_UP, HW_TASK_STATE_IDLE);
+		return NRF_802154_SL_LPTIMER_PLATFORM_TOO_DISTANT;
+	}
+
+	evt_address = z_nrf_rtc_timer_compare_evt_address_get(m_hw_task.chan);
+
+	nrf_802154_sl_mcu_critical_enter(mcu_cs_state);
+
+	/* For triggering to take place a safe margin is 2 lpticks from `now`. */
+	if ((z_nrf_rtc_timer_read() + 2) > fire_lpticks) {
+		/* it is too late */
+		nrf_802154_sl_mcu_critical_exit(mcu_cs_state);
+		cc_unbind(m_hw_task.chan, ppi_channel);
+		m_hw_task.ppi = NRF_802154_SL_HW_TASK_PPI_INVALID;
+		z_nrf_rtc_timer_abort(m_hw_task.chan);
+		done_on_time = false;
+	} else {
+		if (ppi_channel != NRF_802154_SL_HW_TASK_PPI_INVALID) {
+			nrfx_gppi_event_endpoint_setup(ppi_channel, evt_address);
+		}
+		m_hw_task.ppi = ppi_channel;
+		nrf_802154_sl_mcu_critical_exit(mcu_cs_state);
+		done_on_time = true;
+	}
+
+	hw_task_state_set(HW_TASK_STATE_SETTING_UP,
+			  done_on_time ? HW_TASK_STATE_READY : HW_TASK_STATE_IDLE);
+
+	return done_on_time ? NRF_802154_SL_LPTIMER_PLATFORM_SUCCESS :
+	       NRF_802154_SL_LPTIMER_PLATFORM_TOO_LATE;
+}
+
+nrf_802154_sl_lptimer_platform_result_t nrf_802154_platform_sl_lptimer_hw_task_cleanup(void)
+{
+	if (!hw_task_state_set(HW_TASK_STATE_READY, HW_TASK_STATE_CLEANING)) {
+		return NRF_802154_SL_LPTIMER_PLATFORM_WRONG_STATE;
+	}
+
+	z_nrf_rtc_timer_abort(m_hw_task.chan);
+
+	cc_unbind(m_hw_task.chan, m_hw_task.ppi);
+	m_hw_task.ppi = NRF_802154_SL_HW_TASK_PPI_INVALID;
+
+	hw_task_state_set(HW_TASK_STATE_CLEANING, HW_TASK_STATE_IDLE);
+
+	return NRF_802154_SL_LPTIMER_PLATFORM_SUCCESS;
+}
+
+nrf_802154_sl_lptimer_platform_result_t nrf_802154_platform_sl_lptimer_hw_task_update_ppi(
+	uint32_t ppi_channel)
+{
+	bool cc_triggered;
+
+	if (!hw_task_state_set(HW_TASK_STATE_READY, HW_TASK_STATE_UPDATING)) {
+		return NRF_802154_SL_LPTIMER_PLATFORM_WRONG_STATE;
+	}
+
+	cc_bind_to_ppi(m_hw_task.chan, ppi_channel);
+	m_hw_task.ppi = ppi_channel;
+
+	cc_triggered = cc_event_check(m_hw_task.chan);
+
+	hw_task_state_set(HW_TASK_STATE_UPDATING, HW_TASK_STATE_READY);
+
+	return cc_triggered ? NRF_802154_SL_LPTIMER_PLATFORM_TOO_LATE :
```

Comment:
No matter if cc_triggered is true or false, the hw_task state will endup being `HW_TASK_STATE_READY` - shouldn't it be changed to `HW_TASK_STATE_IDLE`?

Reply:
Leaving `HW_TASK_STATE_IDLE` would be a good idea if this function could do a complete cleanup as well. But that's not the idea - it's just an update attempt, so after a temporary state change, the original `HW_TASK_STATE_READY` value is being restored.

Reply:
Ok, but I do not understand what is the use of being in `HW_TASK_STATE_READY` if the timer event has already triggered - it does not seem we can do anything about it.

Of course it is possible, that the reason for that is to remain consequent in API, meaning the user of the API always must call `nrf_802154_platform_sl_lptimer_hw_task_cleanup` explicitly.